src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -1,100 +1,312 @@
+                _projectToInfo = projectToInfo;
+                _metadataPathToInfo = metadataPathToInfo;
             }
 
-            public override Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancellationToken)
+            public override Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, CancellationToken cancellationToken)
             {
-                // check whether we are good to report total symbol numbers
-                if (_symbolCountByProjectMap == null || _symbolCountByProjectMap.Count < solution.ProjectIds.Count || string.IsNullOrEmpty(solution.FilePath))
+                if (!document.SupportsSyntaxTree)
                 {
+                    // Not a language we can produce indices for (i.e. TypeScript).  Bail immediately.
                     return SpecializedTasks.EmptyTask;
                 }
 
-                if (_solutionId != null && _solutionId != solution.Id)
+                if (bodyOpt != null)
                 {
-                    ReportCount();
+                    // This was a method level edit.  This can't change the symbol tree info
+                    // for this project.  Bail immediately.
                     return SpecializedTasks.EmptyTask;
                 }
 
-                _solutionId = solution.Id;
-                foreach (var projectId in solution.ProjectIds)
+                return UpdateSymbolTreeInfoAsync(document.Project, cancellationToken);
+            }
+            
+            public override Task AnalyzeProjectAsync(Project project, bool semanticsChanged, CancellationToken cancellationToken)
+            {
+                return UpdateSymbolTreeInfoAsync(project, cancellationToken);
+            }
+
+            private async Task UpdateSymbolTreeInfoAsync(Project project, CancellationToken cancellationToken)
+            {
+                if (!project.SupportsCompilation)
                 {
-                    if (!_symbolCountByProjectMap.ContainsKey(projectId))
-                    {
-                        return SpecializedTasks.EmptyTask;
-                    }
+                    return;
+                }
+
+                Project lastSeenProject;
+                if (_idTolastSeenProject.TryGetValue(project.Id, out lastSeenProject) && lastSeenProject == project)
+                {
+                    // We already saw this project.  No need to do anything;
+                    return;
                 }
 
-                ReportCount();
-                return SpecializedTasks.EmptyTask;
+                // Check the semantic version of this project.  The semantic version will change
+                // if any of the source files changed, or if the project version itself changed.
+                // (The latter happens when something happens to the project like metadata 
+                // changing on disk).
+                var version = await project.GetSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+
```

Comment:
you already play with version, so I think you can actually save this version and check the saved version with this version to see whether we need to process it or not.